civicrm_entity.module

```diff
@@ -109,10 +115,108 @@ function civicrm_entity_entity_type_build(array &$entity_types) {
         'field_ui_base_route' => "entity.$entity_type_id.collection",
       ]);
     }
+
+    // If this entity has bundle support, we define the bundle field as "bundle"
+    // and will use the "bundle property" as the field to fetch field options
+    // from CiviCRM with.
+    //
+    // @see civicrm_entity_entity_bundle_info()
+    // @see \Drupal\civicrm_entity\Entity\CivicrmEntity::baseFieldDefinitions()
+    if (!empty($civicrm_entity_info['bundle property'])) {
+      $entity_type_info['entity_keys']['bundle'] = 'bundle';
+      $entity_type_info['civicrm_bundle_property'] = $civicrm_entity_info['bundle property'];
+      $entity_type_info['links']['add-page'] = $entity_type_info['links']['add-form'];
+      $entity_type_info['links']['add-form'] = sprintf('%s/{%s}', $entity_type_info['links']['add-form'], $entity_type_info['entity_keys']['bundle']);
+    }
+
     $entity_types[$entity_type_id] = new ContentEntityType($entity_type_info);
   }
 }
 
+/**
+ * Implements hook_entity_bundle_info().
+ */
+function civicrm_entity_entity_bundle_info() {
+  $transliteration = \Drupal::transliteration();
+  /** @var \Drupal\civicrm_entity\CiviCrmApiInterface $civicrm_api */
+  $civicrm_api = \Drupal::service('civicrm_entity.api');
+
+  $bundles = [];
+  $entity_types_with_bundles = array_filter(SupportedEntities::getInfo(), static function(array $civicrm_entity_info) {
+    return !empty($civicrm_entity_info['bundle property']);
+  });
+  foreach ($entity_types_with_bundles as $entity_type_id => $civicrm_entity_info) {
+    // We keep a bundle that is the same as the entity type ID. This allows us
+    // to create fields as if this entity has no bundles.
+    $bundles[$entity_type_id] = [
+      $entity_type_id => [
+        'label' => $civicrm_entity_info['civicrm entity label'],
+      ],
+    ];
```

Comment:
Here is a trick! So we keep the existing bundle as entity type ID along with all of the extra options. Technically nothing will be this value. But it preserves backward compatibility _and_ helps make the user interface work.